mcs/class/corlib/System.Threading/WaitHandle.cs

```diff
@@ -47,6 +47,61 @@ public abstract partial class WaitHandle
 
 		internal const int MaxWaitHandles = 64;
 
+		// We rely on the reference source implementation of WaitHandle, and it delegates to a function named
+		//  WaitOneNative to perform the actual operation of waiting on a handle.
+		// This native operation actually has to call back into managed code and invoke .Wait
+		//  on the current SynchronizationContext. As such, our implementation of this "native" method
+		//  is actually managed code, and the real native icall being used is Wait_internal.
+		// Safe access to Wait_internal is provided by the WaitMultiple and WaitSingle (previously
+		//  called WaitOneNative) methods defined further below.
+		static int WaitOneNative (SafeHandle waitableSafeHandle, uint millisecondsTimeout, bool hasThreadAffinity, bool exitContext)
+		{
+			var context = SynchronizationContext.Current;
+
+			// HACK: Documentation (and public posts by experts like Joe Duffy) suggests that
```

Comment:
Why are we duplicating the code with `WaitSingle`? The existing WaitOneNative could be refactored as follow:
```

// We rely on the reference source implementation of WaitHandle, and it delegates to a function named
//  WaitOneNative to perform the actual operation of waiting on a handle.
// This native operation actually has to call back into managed code and invoke .Wait
//  on the current SynchronizationContext. As such, our implementation of this "native" method
//  is actually managed code, and the real native icall being used is Wait_internal.
static int WaitOneNative (SafeHandle waitableSafeHandle, uint millisecondsTimeout, bool hasThreadAffinity, bool exitContext)
{
	var context = SynchronizationContext.Current;

	bool release = false;
	try {
		waitableSafeHandle.DangerousAddRef(ref release);

#if !DISABLE_REMOTING
		if (exitContext)
			SynchronizationAttribute.ExitContext ();
#endif

		// HACK: Documentation (and public posts by experts like Joe Duffy) suggests that
		//  users must first call SetWaitNotificationRequired to flag that a given synchronization
		//  context overrides .Wait. Because invoking the Wait method is somewhat expensive, we use
		//  the notification-required flag to determine whether or not we should invoke the managed
		//  wait method.
		// Another option would be to check whether this context uses the default Wait implementation,
		//  but I don't know of a cheap way to do this that handles derived types correctly.
		// If the thread does not have a synchronization context set at all, we can safely just
		//  jump directly to invoking Wait_internal.
		if ((context != null) && context.IsWaitNotificationRequired ()) {
			return context.Wait (
				// TODO: This alloc could potentially be optimized out for common cases. Since
				//  waits are a common operation it might be worth it to do that.
				new IntPtr[] { waitableSafeHandle.DangerousGetHandle () },
				// TODO: The value of this shouldn't matter for a single element,
				//  but it's possible code depends on it. Should test what coreclr & windows CLR do
				false, 
				// FIXME: Ugh, there's an implicit uint->int cast here by design...
				(int)millisecondsTimeout
			);
		} else {
			unsafe {
				IntPtr handle = waitableSafeHandle.DangerousGetHandle();
				return Wait_internal(&handle, 1, false, (int)millisecondsTimeout);
			}
		}
	} finally {
		if (release)
			waitableSafeHandle.DangerousRelease ();

#if !DISABLE_REMOTING
		if (exitContext)
			SynchronizationAttribute.EnterContext ();
#endif
	}
}
```

Reply:
Outside code depends on it, so I left it alone as-is. It might be worth it to do this, I'll mess with it...

Reply:
AFAIK outside code depends on `WaitOneNative` and your PR already change the behavior of it. Changing the behavior in this case is fine IMO.